File: frontend/hibit-web/src/app/core/auth/auth.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map, tap } from 'rxjs';
import { environment } from '../../../environments/environment';
import { LoginResponse } from './auth.models';

const TOKEN_KEY = 'hibit_auth_token';
const EXPIRES_KEY = 'hibit_auth_expires_at';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly http = inject(HttpClient);

  getToken(): string | null {
    return sessionStorage.getItem(TOKEN_KEY);
  }

  hasValidToken(): boolean {
    const token = this.getToken();
    const expiresAt = sessionStorage.getItem(EXPIRES_KEY);
    if (!token || !expiresAt) {
      return false;
    }
    return new Date(expiresAt).getTime() > Date.now();
  }

  login(): Observable<void> {
    return this.http
      .post<LoginResponse>(`${environment.apiUrl}/api/auth/login`, {
        username: environment.appUser.username,
        password: environment.appUser.password,
      })
      .pipe(
        tap((response) => this.storeToken(response)),
        map(() => void 0),
      );
  }

  ensureAuthenticated(): Promise<void> {
    if (this.hasValidToken()) {
      return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
      this.login().subscribe({
        next: () => resolve(),
        error: (err) => reject(err),
      });
    });
  }

  clearToken(): void {
    sessionStorage.removeItem(TOKEN_KEY);
    sessionStorage.removeItem(EXPIRES_KEY);
  }

  private storeToken(response: LoginResponse): void {
    sessionStorage.setItem(TOKEN_KEY, response.accessToken);
    sessionStorage.setItem(EXPIRES_KEY, response.expiresAt);
  }
}

File: frontend/hibit-blazor/Services/AuthService.cs
using System.Net.Http.Json;
using Microsoft.JSInterop;
using Microsoft.Extensions.Options;
using Hibit.Web.Models;

namespace Hibit.Web.Services;

public interface IAuthService
{
    Task<bool> EnsureAuthenticatedAsync();
    Task LoginAsync();
    Task<bool> HasValidTokenAsync();
    Task<string?> GetTokenAsync();
    Task ClearTokenAsync();
}

public class AuthService : IAuthService
{
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly IJSRuntime _js;
    private readonly ApiSettings _settings;

    private const string TokenKey = "hibit_auth_token";
    private const string ExpiresKey = "hibit_auth_expires_at";
    private const string AuthClientName = "Auth";

    public AuthService(IHttpClientFactory httpClientFactory, IJSRuntime js, IOptions<ApiSettings> settings)
    {
        _httpClientFactory = httpClientFactory;
        _js = js;
        _settings = settings.Value;
    }

    public async Task<bool> EnsureAuthenticatedAsync()
    {
        if (await HasValidTokenAsync())
        {
            return true;
        }

        await LoginAsync();
        return await HasValidTokenAsync();
    }

    public async Task LoginAsync()
    {
        try
        {
            var http = _httpClientFactory.CreateClient(AuthClientName);
            var request = new LoginRequest
            {
                Username = _settings.AppUser.Username,
                Password = _settings.AppUser.Password
            };

            var response = await http.PostAsJsonAsync("api/auth/login", request);

            if (response.IsSuccessStatusCode)
            {
                var loginResponse = await response.Content.ReadFromJsonAsync<LoginResponse>();
                if (loginResponse != null)
                {
                    await StoreTokenAsync(loginResponse);
                }
            }
        }
        catch
        {
            // API may be down during local development
        }
    }

    public async Task<bool> HasValidTokenAsync()
    {
        var token = await GetTokenAsync();
        var expiresAt = await GetExpiresAtAsync();

        if (string.IsNullOrEmpty(token) || string.IsNullOrEmpty(expiresAt))
        {
            return false;
        }

        if (DateTimeOffset.TryParse(expiresAt, out var expiry))
        {
            return expiry > DateTimeOffset.UtcNow;
        }

        return false;
    }

    public async Task<string?> GetTokenAsync()
    {
        return await _js.InvokeAsync<string?>("sessionStorage.getItem", TokenKey);
    }

    public async Task ClearTokenAsync()
    {
        await _js.InvokeVoidAsync("sessionStorage.removeItem", TokenKey);
        await _js.InvokeVoidAsync("sessionStorage.removeItem", ExpiresKey);
    }

    private async Task StoreTokenAsync(LoginResponse response)
    {
        await _js.InvokeVoidAsync("sessionStorage.setItem", TokenKey, response.AccessToken);
        await _js.InvokeVoidAsync("sessionStorage.setItem", ExpiresKey, response.ExpiresAt);
    }

    private async Task<string?> GetExpiresAtAsync()
    {
        return await _js.InvokeAsync<string?>("sessionStorage.getItem", ExpiresKey);
    }
}
